File: Java/1431. Kids With the Greatest Number of Candies.java
import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode.com/problems/kids-with-the-greatest-number-of-candies/
 */

public class KidsWithTheGreatestNumberOfCandies {
    // O(n) time complexity (2 linear : 2n --> O(n))
    // O(n) space complexity
    List<Boolean> kidsWithCandies(int[] candies, int extraCandies) {
        List<Boolean> res = new ArrayList();
        int max = 0;
        for (int candy : candies)
            max = Math.max(max, candy);
        for (int candy : candies)
            res.add(candy + extraCandies >= max);
        return res;
    }
}

class T {
    public static void main(String[] args) {
        KidsWithTheGreatestNumberOfCandies k = new KidsWithTheGreatestNumberOfCandies();
        int[] candies = {2, 3, 5, 1, 3};
        int extraCandies = 3;
        List<Boolean> res = k.kidsWithCandies(candies, extraCandies);
        for (Boolean v : res) {
            System.out.print(v + " "); // true,true,true,false,true
        }
    }
}

File: 1431. Kids With the Greatest Number of Candies.kt
/**
 * https://leetcode.com/problems/running-sum-of-1d-array/
 */
// O(n) time complexity
// O(1) space complexity
fun kidsWithCandies(candies: IntArray, extraCandies: Int): BooleanArray {
    val max = candies.max()
    val res = BooleanArray(candies.size)
    for(i in candies.indices)
        res[i] = candies[i]+extraCandies >= max!!
    return res
}

fun main() {
    val candies = intArrayOf(2, 3, 5, 1, 3)
    val extraCandies = 3
    val res = kidsWithCandies(candies, extraCandies)
    for (v in res) {
        print("$v ") // true,true,true,false,true
    }
}